persist/file-adapter/adapter.go

```diff
@@ -113,11 +113,29 @@ func (a *Adapter) AddPolicy(sec string, ptype string, rule []string) error {
 	return errors.New("not implemented")
 }
 
+// AddPolicies adds policy rules to the storage.
+func (a *Adapter) AddPolicies(sec string, ptype string, rules [][]string) []error {
```

Comment:
We should only return one error, similar for others.

If one add failed, we should roll back previous ones. So it's atomic operation.

Reply:
Rollback should not exist. It increases complexity. We just returns an error is better.

Reply:
> We should only return one error, similar for others.
> 
> If one add failed, we should roll back previous ones. So it's atomic operation.

To add atomicity(add or remove everything or nothing) to the operations, it occurs to me, there are two solutions to tackle this :
1. We add or remove the policy rule one by one and when we get an invalid policy rule we remove or add the previously added or removed policy rules respectively.
OR
2. Introduce the function which checks the validity of the policy rules if all the policy rules are valid then we move forward for the operation or we do not and pass the error.

I think the 2nd solution is better, what do you think?

Reply:
> Rollback should not exist. It increases complexity. We just returns an error is better.

Then how do you manage to fix for previously added or removed rules from the policy?

Reply:
And also we can give support for two functions, like ```AddPolicies()```, in this case, the return type is ```[]error``` and ```AddAtomicPolicies()```, in this case, the return type is ```error```. This can be fruitful for the user.

Reply:
What do you think, @hsluoyz and @nodece ?

Reply:
> > We should only return one error, similar for others.
> > If one add failed, we should roll back previous ones. So it's atomic operation.
> 
> To add atomicity(add or remove everything or nothing) to the operations, it occurs to me, there are two solutions to tackle this :
> 
> 1. We add or remove the policy rule one by one and when we get an invalid policy rule we remove or add the previously added or removed policy rules respectively.
>    OR
> 2. Introduce the function which checks the validity of the policy rules if all the policy rules are valid then we move forward for the operation or we do not and pass the error.
> 
> I think the 2nd solution is better, what do you think?

I agree it.

Reply:
Okay